umapi_client/connection.py

```diff
@@ -310,6 +313,20 @@ def execute_multiple(self, actions, immediate=True):
         for a in actions:
             if len(a.commands) == 0:
                 if self.logger: self.logger.warning("Sending action with no commands: %s", a.frame)
+            # throttling part 1a: split commands with group assignments > self.throttle_groups
```

Comment:
So there are several issues here:

- since the logic for finding overly-long group lists requires iterating over the list, splitting the logic for checking whether the command needs to be split from the logic for actually splitting the command leads to a really tortured definition for `split_groups` where you are pointing it at a particular dictionary entry in an indexed command.
- you are modifying a list that you are iterating over, which is a very dangerous and non-Pythonic practice that almost for sure is leading to unexpected results (e.g., I suspect you may either be re-examining the new commands you split out from the rest or not examining commands that get pushed out by the insertion of the new commands).  Never do this.
- you are not logging whether you made any changes in the structure of the action, which is really the only excuse for putting a test for whether the splitting is necessary anyway.

I strongly recommend that you instead alter your split_groups command to take *only* the `max_groups` parameter, rename it `maybe_split_groups`, and have it return whether or not it split any commands.  Then you can integrate the testing for long groups in with the fixing of them, leading to a single iteration happening in that one function (which you need to do a lot more carefully than you are here).  And the code here becomes:
```
if a.maybe_split_groups(self.throttle_groups):
    if self.logger: self.logger.debug("Throttling actions %s to have a maximum of %d entries in group lists.",
                                                           a.frame, self.throttle_groups)
```
and then you can integrate this comment with the one above, so it says "throttling groups and commands".